### PR TITLE
Add telemetry event for HostGAPlugin retry logic

### PR DESCRIPTION
**Pull request by pgombar**
## Description

We want to emit telemetry whenever we invoke the HostGAPlugin retry logic and log the success or failure of the calls (first or second try). For that, `WireProtocol` needed to use the `add_event` method, which introduced circular dependencies.

Imports are fixed with the following changes:
- Extract the `DataContract` object(s) and related methods to a common module outside of `restapi`.
- Extract the `TelemetryEvent` object(s) to a common module outside of `restapi`.
- Update all imports referencing these objects and methods.

Also, added sleep mocking in `test_extension` module to reduce unit test execution time from 7+ minutes to ~2.5 minutes.

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1632)

### CODE REVIEW

File: azurelinuxagent/common/datacontract.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+from azurelinuxagent.common.exception import ProtocolError
+import azurelinuxagent.common.logger as logger
+
+"""
+Data contract between guest and host
```
Comment (narrieta) [inline azurelinuxagent/common/datacontract.py]: "Base class for data contracts between guest and host and utilities to manipulate the properties in those contracts"

File: azurelinuxagent/common/datacontract.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+"""
```
Comment (narrieta) [inline azurelinuxagent/common/datacontract.py]: There some imports referencing the old module, please do a global search of usages of these classes/methods.

E.g.
> from azurelinuxagent.common.protocol.restapi import DataContract, set_properties

in imds.py

File: azurelinuxagent/common/protocol/metadata.py [hunk 3]
```diff
@@ -25,9 +25,14 @@
 import traceback
 
 import azurelinuxagent.common.conf as conf
+from azurelinuxagent.common.datacontract import get_properties, set_properties, validate_param
+from azurelinuxagent.common.exception import HttpError, ProtocolError
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/metadata.py]: what was happening here? I see metadata.py does reference HttpError, etc... but there was no import before???

Comment (narrieta) [reply azurelinuxagent/common/protocol/metadata.py]: was working because of the import * from restapi at line 38?

File: tests/ga/test_extension.py [hunk 4]
```diff
@@ -27,6 +27,12 @@
 from azurelinuxagent.ga.exthandlers import *
 from azurelinuxagent.common.protocol.wire import WireProtocol, InVMArtifactsProfile
 
+SLEEP = time.sleep
```
Comment (narrieta) [inline tests/ga/test_extension.py]: Let's add some comments here

File: azurelinuxagent/common/protocol/metadata.py (hunk 3 above)
Comment (larohra) [reply azurelinuxagent/common/protocol/metadata.py]: restapi was importing this `from azurelinuxagent.common.exception import ProtocolError, HttpError` before and the `import *` was pulling it in

File: azurelinuxagent/common/protocol/wire.py [hunk 5]
```diff
@@ -1024,18 +1031,45 @@ def send_request_using_appropriate_channel(self, direct_func, host_func):
...
+                              log_event=False)
+
+            except Exception as e:
+                add_event(name=AGENT_NAME,
+                          version=CURRENT_VERSION,
+                          op=WALAEventOperation.HostPlugin,
+                          is_success=False,
+                          message=ustr(e),
+                          log_event=False)
+                raise e
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py]: "raise" instead of "raise e" to re-raise the original exception

File: azurelinuxagent/common/protocol/wire.py [hunk 6]
```diff
@@ -1024,18 +1031,45 @@ def send_request_using_appropriate_channel(self, direct_func, host_func):
...
+
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py]: Why is this needed?

```
         except Exception:
            raise
```

File: azurelinuxagent/common/protocol/metadata.py (hunk 3 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/metadata.py]: yup --- hidden/unintended dependencies --- please stay away from this kind of imports in new code (and let's try to fix existing code when possible)

File: azurelinuxagent/common/protocol/wire.py [hunk 7]
```diff
@@ -1024,18 +1031,45 @@ def send_request_using_appropriate_channel(self, direct_func, host_func):
...
+                                                                             old_role_config_name, new_role_config_name)
+                    add_event(name=AGENT_NAME,
+                              version=CURRENT_VERSION,
+                              op=WALAEventOperation.HostPlugin,
+                              is_success=True,
+                              message=msg,
+                              log_event=False)
+
+            except Exception as e:
+                add_event(name=AGENT_NAME,
```
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/common/protocol/wire.py]: why restrict the error to only ResourceGone and InvalidContainer? Shouldn't we log that we failed irrespective of the error type?

[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: the intention is to measure whether the refresh helped to solve ResourceGone/InvalidContainer
if retry fails with ResourceGone/InvalidContainer then it didn't help; if it fails with some other error then it's something else

[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: > if it fails with some other error then it's something else
you may want to measure whether the retry introduced another issue, but i wouldn't go there

[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/common/protocol/wire.py]: I agree to the measurement but we should filter out ResourceGone/InvalidContainer in the query rather than out here in the source. Logging all error might help us investigate unknown issues later on

[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: if you want to measure failures communicating with the hostgaplugin then we need more instrumentation (e.g. failures in the first call, not only on the retry)

[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: btw, I am thinking these measurements should be only temporary --- we are measuring the effectiveness of specific fixes, and we may end up accummulating tons of those over time
we should have an offline talk about this

File: azurelinuxagent/common/protocol/metadata.py (hunk 3 above)
Comment (pgombar) [reply azurelinuxagent/common/protocol/metadata.py]: Exactly, my reasoning was to eliminate all the `*` imports in the files in was changing. It's scary.

File: azurelinuxagent/common/protocol/wire.py (hunk 5 above)
Comment (pgombar) [reply azurelinuxagent/common/protocol/wire.py]: Thanks

File: azurelinuxagent/common/protocol/wire.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pgombar) [reply azurelinuxagent/common/protocol/wire.py]: Agree with Norberto. Since the purpose of this event is to merely gather data about the fix for `ResourceGoneError` and `InvalidContainerError`, it makes sense to only emit events for those exceptions.

File: azurelinuxagent/common/protocol/wire.py (hunk 6 above)
Comment (pgombar) [reply azurelinuxagent/common/protocol/wire.py]: I needed a try/catch to emit a telemetry event, but originally we weren't swallowing the exceptions from the retry call.

File: azurelinuxagent/common/datacontract.py (hunk 1 above)
Comment (pgombar) [reply azurelinuxagent/common/datacontract.py]: Thanks

File: tests/ga/test_extension.py (hunk 4 above)
Comment (pgombar) [reply tests/ga/test_extension.py]: Added

File: azurelinuxagent/common/protocol/wire.py (hunk 6 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py]: But it is not needed. We aren't doing anything in the exception handler, just re-raising the exception.

File: azurelinuxagent/common/protocol/wire.py [hunk 8]
```diff
@@ -1024,22 +1031,58 @@ def send_request_using_appropriate_channel(self, direct_func, host_func):
...
+            try:
+                ret = host_func()
+                if ret:
+                    msg = "Request succeeded using the host plugin channel after goal state refresh." \
+                          "ContainerId changed from {0} to {1}, " \
+                          "role config file changed from {2} to {3}.".format(old_container_id, new_container_id,
+                                                                             old_role_config_name, new_role_config_name)
+                    add_event(name=AGENT_NAME,
+                              version=CURRENT_VERSION,
+                              op=WALAEventOperation.HostPlugin,
```
Comment (larohra) [inline azurelinuxagent/common/protocol/wire.py]: Are we gonna report the event always? Or is it only for monitoring this release?

Comment (pgombar) [reply azurelinuxagent/common/protocol/wire.py]: I'm going for always reporting. We were missing pieces of info before related to HostGAPlugin, with these events we'd have a more complete picture when investigating issues.

File: azurelinuxagent/common/protocol/wire.py [hunk 9]
```diff
@@ -1024,22 +1031,58 @@ def send_request_using_appropriate_channel(self, direct_func, host_func):
...
+                              is_success=True,
+                              message=msg,
+                              log_event=False)
+
+            except (ResourceGoneError, InvalidContainerError) as e:
+                add_event(name=AGENT_NAME,
+                          version=CURRENT_VERSION,
+                          op=WALAEventOperation.HostPlugin,
+                          is_success=False,
+                          message=ustr(e),
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/wire.py]: The message should be more explicit, something along the lines of the message at line 1059.
Also, probably we should not include the entire exception, but only its type (in case there are any secrets in the request)

Comment (pgombar) [reply azurelinuxagent/common/protocol/wire.py]: Good catch, thanks. Updated.
